File: 2024/src/day01/day01.py
from collections import Counter


def processing() -> tuple[list[int], list[int]]:
    lines: list[list[int]] = [list(map(int, line.split())) for line in open("src/day01/input.txt")]
    return [lines[i][1] for i in range(len(lines))], [lines[i][0] for i in range(len(lines))]


def part1() -> int:
    left, right = map(sorted, processing())
    return sum([abs(left[i] - right[i]) for i in range(len(left))])


def part2() -> int:
    # noinspection PyTypeChecker
    left, right = map(Counter, processing())
    return sum([0 if key not in right else key * value * right[key] for key, value in left.items()])


if __name__ == "__main__":
    print(part1())
    print(part2())

File: 2023/src/day01/day01.go
package main

import (
	"2023/src/utils"
	"fmt"
	"strings"
)

func processing(filename string) []string {
	text := utils.ReadFile(filename)
	lines := strings.SplitN(text, "\n", -1)
	return lines
}

func calc(line string) int {
	return 10*int(line[strings.IndexAny(line, "0123456789")]-'0') + int(line[strings.LastIndexAny(line, "0123456789")]-'0')
}

func part1(filename string) int {
	var result int
	lines := processing(filename)
	for _, line := range lines {
		result += calc(line)
	}
	return result
}

func part2(filename string) int {
	var result int
	lines := processing(filename)
	replacer := strings.NewReplacer("one", "o1e", "two", "t2o", "three", "t3e", "four", "f4r", "five", "f5e", "six", "s6x", "seven", "s7n", "eight", "e8t", "nine", "n9e", "zero", "z0o")
	for _, line := range lines {
		result += calc(replacer.Replace(replacer.Replace(line)))
	}
	return result
}

func main() {
	filename := "src/day01/input.txt"
	fmt.Println(part1(filename))
	fmt.Println(part2(filename))
}
